base/compiler/ssair/passes.jl

```diff
@@ -1880,6 +1880,196 @@ function adce_pass!(ir::IRCode, inlining::Union{Nothing,InliningState}=nothing)
+        # copyto!(key, stmt.args)
+        # key[end] = stmt.head
+        # for (i, arg) in enumerate(stmt.args)
+        #     if isa(arg, SSAValue)
+        #         key[i] = SSAValue(ssa_to_ssa[arg.id])
+        #     end
+        # end
+        # svec(key...)
+
+        key = ccall(:jl_alloc_svec, SimpleVector, (Csize_t,), length(stmt.args)+1);
+        ptr = convert(Ptr{Ptr{Cvoid}}, pointer_from_objref(key)) + sizeof(Ptr{Cvoid})
+        for (i, arg) in enumerate(stmt.args)
+            if isa(arg, SSAValue)
+                unsafe_store!(ptr, ccall(:jl_box_ssavalue, Ptr{Any}, (Csize_t,), ssa_to_ssa[arg.id]), i)
+            else
+                unsafe_store!(ptr, arrayptr(Ptr{Nothing}, stmt.args, i), i)
+            end
+        end
+        unsafe_store!(ptr, pointer_from_objref(stmt.head), length(stmt.args)+1)
+        return key
+    else
+        return nothing
+    end
+end
+function perform_symbolic_evaluation(stmt::PhiNode, ssa_to_ssa, blockidx)
+    length(stmt.values) == 0 && return nothing
+
+    no_of_edges = length(stmt.edges)
+    key = Vector{Any}(undef, no_of_edges*2 + 2)
+    copyto!(key, stmt.edges)
+    key[no_of_edges+1] = :phi
+    key[no_of_edges+2] = blockidx
+    copyto!(key, no_of_edges+3, stmt.values)
+
+    firstval = nothing
+    allthesame = true
+    deletions = 0
+    for j in eachindex(stmt.values)
+        !isassigned(stmt.values, j) && return nothing
+
+        val = stmt.values[j]
+        if val isa SSAValue && ssa_to_ssa[val.id] == 0
+            deleteat!(key, j-deletions)
+            deletions += 1
+            deleteat!(key, no_of_edges+2-deletions+j)
+            deletions += 1
+        else
+            if val isa SSAValue
+                key[no_of_edges+2-deletions+j] = ssa_to_ssa[val.id]
+            end
+            if firstval === nothing
+                firstval = val
+            else
+                allthesame &= val === firstval
+            end
+        end
+    end
+    return svec(key...)
+end
+
+"""
+    gnv!(ir::IRCode) -> newir::IRCode
+
+Global Value Numbering (GVN) pass.
+
+This is based on the NewGVN pass in LLVM.
+
+GVN partitions all the statements in the IR into congruence classes. All elements in a congruence class are guaranteed to be the same.
+
+It implements the RPO value numbering algorithm based on the paper "SCC based value numbering" by L. Taylor Simpson.
+The elimination step is based on the implemenation in LLVM's NewGVN pass.
+"""
+function gvn!(ir::IRCode)
+    changed = true
+    ssa_to_ssa = fill(0, length(ir.stmts.stmt)) # Map from ssa to ssa of equivalent value
+    # Value type is SSAValue in order to reuse cache from it being boxed in svec
+    val_to_ssa = IdDict{SimpleVector, SSAValue}() # Map from value of an expression to ssa with equivalent value
+    sizehint!(val_to_ssa, length(ir.stmts.stmt))
+
+    while changed
+        changed = false
+
+        # Reverse Post Order traversal of dominator tree
```

Comment:
Comment should mention why this is an RPO traversal (I.e. it's an IR invariant).